File: src/main/scala/com/hmrc/scart/scala/LineItem.scala
/**
 *
 */
package com.hmrc.scart.scala


import com.hmrc.scart.constants.SCartConstants
import org.json.JSONObject

/**
 * @author skotha
 *
 */
class LineItem(val sku : JSONObject, val skuName : String, val quantity : Int, val unitPrice : BigDecimal, val basePrice:BigDecimal,val discountAmount : BigDecimal,
    val yourPrice :BigDecimal, val currency :String) {
   
  
  def updateQuantity(qty : Int) : LineItem = {
    val quantity = LineItem.qanty(qty) 
    val basePrice = (unitPrice.*(quantity))
     val discount = LineItem.calculateDiscount(sku, quantity)
     new LineItem(sku, skuName,quantity, unitPrice, basePrice, discount, (basePrice.-(discount)), currency)};
     
  def toJsonObject : JSONObject  =
  {
     new JSONObject().put(SCartConstants.SKU_NAME, skuName). put(SCartConstants.QTY, quantity). put(SCartConstants.UNIT_PRICE, unitPrice).
        put(SCartConstants.BASE_PRICE, basePrice). put(SCartConstants.CURRENCY, currency). put(SCartConstants.YOU_SAVE, discountAmount).
        put(SCartConstants.YOUR_PRICE, yourPrice);
  } 
  
  
}

object LineItem{
  
  def apply(sku : JSONObject) : LineItem = 
  {
    val quantity : Int = qanty(sku.getInt(SCartConstants.QTY)); 
    val unitPrice : BigDecimal = BigDecimal.apply(sku.getString(SCartConstants.UNIT_PRICE));
    val basePrice : BigDecimal = unitPrice.*(quantity)
    val discount : BigDecimal = calculateDiscount(sku, quantity)

    new LineItem(sku, sku.getString(SCartConstants.SKU_NAME), quantity, unitPrice, 
         basePrice, discount, (basePrice - discount),sku.getString(SCartConstants.CURRENCY))      
  }
  
  def calculateDiscount(sku : JSONObject, quantity : Int) : BigDecimal = {
     val directive : JSONObject = if(sku.has(SCartConstants.DIRECTIVE))sku.getJSONObject(SCartConstants.DIRECTIVE) else null;
    if(directive != null)  
      (quantity / directive.getInt(SCartConstants.FOR_EACH)).*(BigDecimal.apply(directive.getString(SCartConstants.ABSOLUTE_DISCOUNT))) 
    else
           BigDecimal.apply(0);     
   
    }
  
  def qanty(qty : Int) : Int = if(qty < 0) 0 else qty; 
}  

File: src/main/java/com/hmrc/scart/checkout/LineItem.java
/**
 * 
 */
package com.hmrc.scart.checkout;

import java.math.BigDecimal;

import javax.json.Json;
import javax.json.JsonObject;

import com.hmrc.scart.constants.SCartConstants;

/**
 * @author skotha
 *
 */
public class LineItem {

	private final String skuName;
	private final int quantity;
	private final BigDecimal unitPrice;
	private final BigDecimal basePrice;
	private final BigDecimal discountAmount;
	private final BigDecimal yourPrice;
	private final String currency;
	private final JsonObject sku;
	
	public LineItem(JsonObject sku)
	{
		this(sku, sku.getString(SCartConstants.SKU_NAME), sku.getInt(SCartConstants.QTY), 
				new BigDecimal(sku.getString(SCartConstants.UNIT_PRICE)), sku.getString(SCartConstants.CURRENCY));		
	}
	
	private LineItem(JsonObject sku, String skuName, int quantity, BigDecimal unitPrice, String currency)
	{
		this.sku = sku;
		this.skuName = skuName;		
		this.quantity = quantity < 0 ? 0 : quantity;
		this.unitPrice = unitPrice;
		this.basePrice = (this.unitPrice.multiply(new BigDecimal(this.quantity)));
		this.discountAmount = calculateDiscount(this.sku);
		this.yourPrice = this.basePrice.subtract(this.discountAmount);	
		this.currency = currency;		
	}	
	
	public BigDecimal getUnitPrice() {
		return unitPrice;
	}

	public String getSkuName() {
		return skuName;
	}

	public int getQuantity() {
		return quantity;
	}

	public BigDecimal getBasePrice() {
		return basePrice;
	}

	public BigDecimal getDiscountAmount() {
		return discountAmount;
	}

	public BigDecimal getYourPrice() {
		return yourPrice;
	}
	
	public String getCurrency() {
		return currency;
	}
	
	
	public JsonObject getSku() {
		return sku;
	}

	public LineItem updateQuantity(int quantity) {
		return new LineItem(this.sku, this.skuName, quantity, this.unitPrice, this.currency);
	}
	
	private BigDecimal calculateDiscount(JsonObject sku)
	{
		final JsonObject directive = sku.getJsonObject(SCartConstants.DIRECTIVE);
		return directive != null ? new BigDecimal(this.quantity / directive.getInt(SCartConstants.FOR_EACH)).multiply(
				new BigDecimal(directive.getString(SCartConstants.ABSOLUTE_DISCOUNT))) : 
					 BigDecimal.ZERO;
	}
	
	public JsonObject toJsonObject()
	{
		return Json.createObjectBuilder().
				add(SCartConstants.SKU_NAME, getSkuName()).
				add(SCartConstants.QTY, getQuantity()).
				add(SCartConstants.UNIT_PRICE, getUnitPrice()).
				add(SCartConstants.BASE_PRICE, getBasePrice()).
				add(SCartConstants.CURRENCY, getCurrency()).
				add(SCartConstants.YOU_SAVE, getDiscountAmount()).
				add(SCartConstants.YOUR_PRICE, getYourPrice()).build();
	}	
	
}
